presto-main/src/main/java/io/prestosql/sql/planner/optimizations/QueryCardinalityUtil.java

```diff
@@ -143,13 +144,23 @@ public CardinalityExtractorPlanVisitor(Lookup lookup)
         @Override
         public Range<Long> visitLimit(LimitNode node, Void context)
         {
-            Range<Long> sourceCardinalityRange = node.getSource().accept(this, null);
-            long upper = node.getCount();
+            return applyLimit(node.getSource(), node.getCount());
+        }
+
+        @Override
+        public Range<Long> visitTopN(TopNNode node, Void context)
+        {
+            return applyLimit(node.getSource(), node.getCount());
+        }
+
+        private Range<Long> applyLimit(PlanNode source, long limit)
```

Comment:
nit: `s/applyLimit/visitLimitSource` ?

Reply:
> nit: s/applyLimit/visitLimitSource ?

I don't think we should rename it to visitXXX since it makes it easy to confuse with the methods that are implementations of the visitor interface. This is just an internal utility method.